TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/http/retries/DefaultRetry.java

```diff
@@ -40,15 +44,48 @@ public static boolean retryIsUnnecessaryGivenStatusCode(int statusCode) {
+    shouldRetry(ByteBuf targetRequestBytes,
+                List<AggregatedRawResponse> previousResponses,
+                AggregatedRawResponse currentResponse,
+                TrackedFuture<String, ? extends IRequestResponsePacketPair> reconstructedSourceTransactionFuture) {
+        if (previousResponses.size() >= MAX_RETRIES) {
+            return TextTrackedFuture.completedFuture(RequestSenderOrchestrator.RetryDirective.DONE, () -> "done");
+        }
+        return shouldRetry(targetRequestBytes, currentResponse, reconstructedSourceTransactionFuture);
+    }
+
+    public TrackedFuture<String, RequestSenderOrchestrator.RetryDirective>
+    shouldRetry(ByteBuf targetRequestBytes,
+                AggregatedRawResponse currentResponse,
+                TrackedFuture<String, ? extends IRequestResponsePacketPair> reconstructedSourceTransactionFuture) {
         var rr = currentResponse.getRawResponse();
-        if (rr != null && retryIsUnnecessaryGivenStatusCode(rr.status().code())) {
-            return makeDeterminationFuture(RequestSenderOrchestrator.RetryDirective.DONE,
-                "returning DONE because response code was terminal");
+        if (rr != null) {
+            if (retryIsUnnecessaryGivenStatusCode(rr.status().code())) {
+                return makeDeterminationFuture(RequestSenderOrchestrator.RetryDirective.DONE,
+                    "returning DONE because response code was terminal");
+            } else {
+                return reconstructedSourceTransactionFuture.thenCompose(rrp ->
+                        TextTrackedFuture.completedFuture(
+                            Optional.ofNullable(rrp.getResponseData())
+                                .flatMap(sourceResponse ->
+                                    Optional.ofNullable(HttpByteBufFormatter.processHttpMessageFromBufs(
+                                        HttpByteBufFormatter.HttpMessageType.RESPONSE,
+                                        Stream.of(sourceResponse.asByteBuf()))))
+                                .filter(o -> o instanceof HttpResponse)
+                                .map(responseMsg -> shouldRetry(((HttpResponse)responseMsg).status().code(),
+                                    rr.status().code()))
+                                .orElse(RequestSenderOrchestrator.RetryDirective.RETRY),
+                            () -> "evaluating retry status dependent upon source error field"),
+                    () -> "checking the accumulated source response value");
+            }
+        } else {
+            return TextTrackedFuture.completedFuture(RequestSenderOrchestrator.RetryDirective.RETRY, () -> "retrying");
         }
-        return getRetryDirectiveUnlessExceededMaxRetries(previousResponses);
+    }
+
+    private RequestSenderOrchestrator.RetryDirective shouldRetry(int sourceStatus, int targetStatus) {
+        return targetStatus >= 300 && sourceStatus < 300 ?
```

Comment:
A case that's interesting to me here is if the source is 404 and target is 500, we won't retry. This is probably fine from a side effect perspective, but we should at least document this from a test case and written documentation